src/megatron/bridge/recipes/qwen_vl/qwen25_vl_dataset.py

```diff
@@ -0,0 +1,231 @@
+        return input_ids, pixel_values_t, image_grid_thw_t
+
+    def __getitem__(self, idx: int) -> Dict[str, torch.Tensor]:  # noqa: ARG002 - idx unused randomness ok
+        input_ids, pixel_values, image_grid_thw = self._build_inputs()
+
+        # Ensure at least length 2 to form tokens/labels by shifting
+        if input_ids.numel() < 2:
+            # In rare cases of extremely short templates, pad minimally
+            pad_id = self._hf_tokenizer.pad_token_id or 0
+            input_ids = torch.nn.functional.pad(input_ids, (0, 2 - input_ids.numel()), value=pad_id)
+
+        # Create tokens and labels with next-token prediction convention
+        tokens = input_ids[:-1].contiguous()
+        labels = input_ids[1:].contiguous()
+
+        # Position IDs: [0, 1, ..., L-1]
+        position_ids = torch.arange(tokens.numel(), dtype=torch.long, device=tokens.device)
+
+        # Attention mask: 1D valid-token mask (all ones by default)
+        attention_mask = torch.ones_like(tokens, dtype=torch.bool)
+
+        # Loss mask: mask out padding positions in labels
+        pad_token_id = self._hf_tokenizer.pad_token_id
+        if pad_token_id is None:
+            pad_token_id = getattr(self._hf_tokenizer, "eos_token_id", 0) or 0
+
+        loss_mask = torch.ones_like(labels, dtype=torch.float)
+        # Additionally mask special pad token ids if present
+        loss_mask[labels == pad_token_id] = 0.0
+
+        # For embedding lookup safety on padding
+        tokens = tokens.clone()
+        labels = labels.clone()
+        tokens[tokens == pad_token_id] = 0
+        labels[labels == pad_token_id] = 0
+
+        sample: Dict[str, torch.Tensor] = {
+            "tokens": tokens,
+            "labels": labels,
+            "attention_mask": attention_mask,
+            "loss_mask": loss_mask,
+            "position_ids": position_ids,
+        }
+        if pixel_values is not None:
+            sample["pixel_values"] = pixel_values
+        if image_grid_thw is not None:
+            sample["image_grid_thw"] = image_grid_thw
+
+        return sample
+
+
+@dataclass(kw_only=True)
+class MockQwen25VLDatasetProvider(DatasetProvider):
+    """DatasetProvider for a mock Qwen2.5-VL vision-language dataset.
+
+    Builds train/valid/test datasets using a HF AutoProcessor and the
+    MockQwen25VLDataset implementation.
+    """
+
+    # Required to match model.seq_length
+    sequence_length: int
+
+    # HF processor/model ID for Qwen2.5-VL
+    hf_model_path: str = "Qwen/Qwen2.5-VL-3B-Instruct"
+
+    # Sample generation options
+    prompt: str = "Describe this image."
+    random_seed: int = 0
+    image_size: Tuple[int, int] = (256, 256)
+    pad_to_max_length: bool = True
+    create_attention_mask: bool = True
+
+    # Keep parity with GPTDatasetConfig usage in batching utilities
+    skip_getting_attention_mask_from_dataset: bool = True
+
+    # HF AutoProcessor instance will be set during build
+    processor: Optional[Any] = None
```

Comment:
this is not documented right now, but private fields are skipped when saving a checkpoint and serializing the config to yaml. so any fields you don't want serialized in the `run_config.yaml` can be prefixed with `_` 

https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/c7f081bd841bd689cb5a5f46b72f4e98df9ceccd/src/megatron/hub/training/utils/config_utils.py#L121-L122